feature/s3/transfermanager/api_op_GetObject.go

```diff
@@ -0,0 +1,1106 @@
+//
+// Before calling GetObject to download object, you must create a ConcurrentReader and use that reader to
+// copy response content to your final destination file or buffer. This new reader type implements io.Reader to
+// concurrently download parts of large object while limiting the max local cache size during download to prevent
+// too much memory space consumption when getting large objects up to multi-gigabytes. You could configure that buffer
+// size by changing Options.GetBufferSize.
+//
+// Example of creating ConcurrentReader to call GetObject:
+//
+//	file, err := os.Create("your filename")
+//	if err != nil {
+//		log.Fatal("error when creating local file: ", err)
+//	}
+//	r := transfermanager.NewConcurrentReader()
+//	var wg sync.WaitGroup
+//	wg.Add(1)
+//
+// // You must read from the r in a separate goroutine to drive getter to get all parts.
+//
+//	go func() {
+//		defer wg.Done()
+//		_, err := io.Copy(file, r)
+//		if err != nil {
+//			log.Fatal("error when writing to local file: ", err)
+//		}
+//	}()
+//
+//	out, err := svc.GetObject(context.Background(), &transfermanager.GetObjectInput{
+//		Bucket: "your-bucket",
+//		Key:    "your-key",
+//		Reader: r,
+//	})
+//
+//	// must wait for r.Read() to finish
+//	wg.Wait()
+//	if err != nil {
+//		log.Fatal("error when downloading file: ", err)
+//	}
+func (c *Client) GetObject(ctx context.Context, input *GetObjectInput, opts ...func(*Options)) (*GetObjectOutput, error) {
+	i := getter{in: input, options: c.options.Copy(), r: input.Reader}
+	for _, opt := range opts {
+		opt(&i.options)
+	}
+
+	return i.get(ctx)
+}
+
+type getter struct {
+	options Options
+	in      *GetObjectInput
+	out     *GetObjectOutput
+	w       *types.WriteAtBuffer
+	r       *ConcurrentReader
+
+	wg sync.WaitGroup
+	m  sync.Mutex
+
+	offset     int64
+	pos        int64
+	totalBytes int64
+	written    int64
+
+	err error
+}
+
+func (g *getter) get(ctx context.Context) (out *GetObjectOutput, err error) {
+	if err := g.init(ctx); err != nil {
+		return nil, fmt.Errorf("unable to initialize download: %w", err)
+	}
+
+	clientOptions := []func(*s3.Options){
+		func(o *s3.Options) {
+			o.APIOptions = append(o.APIOptions,
+				middleware.AddSDKAgentKey(middleware.FeatureMetadata, userAgentKey),
+				addFeatureUserAgent,
+			)
+		}}
+
+	defer close(g.r.ch)
+	if g.in.PartNumber > 0 {
+		return g.singleDownload(ctx, clientOptions...)
+	}
+
+	if g.options.GetObjectType == types.GetObjectParts {
+		if g.in.Range != "" {
+			return g.singleDownload(ctx, clientOptions...)
+		}
+		// must know the part size before creating stream reader
+		out, err := g.options.S3.HeadObject(ctx, &s3.HeadObjectInput{
+			Bucket:     aws.String(g.in.Bucket),
+			Key:        aws.String(g.in.Key),
+			PartNumber: aws.Int32(1),
+		}, clientOptions...)
+		if err != nil {
+			g.r.setErr(err)
+			return nil, err
+		}
+
+		partsCount := int32(math.Max(float64(aws.ToInt32(out.PartsCount)), 1))
```

Comment:
Don't use math.Max for this, it only supports float64 (hence why you had to do the conversion). It's more idiomatic to just define a local `max()` and use that, we do it in various places.

Reply:
since Go 1.21 there's also a built-in `max` function [ref](https://go.dev/ref/spec#Min_and_max)